kanidmd/src/lib/be/idl_arc_sqlite.rs

```diff
@@ -710,6 +716,257 @@ impl<'a> IdlArcSqliteWriteTransaction<'a> {
         })
     }
 
+    pub fn is_idx_slopeyness_generated(
+        &self,
+        audit: &mut AuditScope,
+    ) -> Result<bool, OperationError> {
+        self.db.is_idx_slopeyness_generated(audit)
+    }
+
+    pub fn get_idx_slope(
+        &self,
+        audit: &mut AuditScope,
+        ikey: &IdxKey,
+    ) -> Result<Option<IdxSlope>, OperationError> {
+        self.db.get_idx_slope(audit, ikey)
+    }
+
+    /// Index Slope Analysis. For the purpose of external modules you can consider this as a
+    /// module that generates "weights" for each index that we have. Smaller values are faster
+    /// indexes - larger values are more costly ones. This is not intended to yield perfect
+    /// weights. The intent is to seperate over obviously more effective indexes rather than
+    /// to min-max the fine tuning of these. Consider name=foo vs class=*. name=foo will always
+    /// be better than class=*, but comparing name=foo to spn=foo is "much over muchness" since
+    /// both are really fast.
+    pub fn analyse_idx_slopes(&mut self, audit: &mut AuditScope) -> Result<(), OperationError> {
+        /*
+         * Inside of this analysis there are two major factors we need to understand
+         *
+         * * What is the variation of idl lengths within an index?
+         * * How man keys are stored in this index?
+         *
+         * Since we have the filter2idl threshold, we want to find "what is the smallest
+         * and most unique index asap so we can exit faster". This allows us to avoid
+         * loading larger most costly indexs that either have large idls, high variation
+         * or few keys and are likely to miss and have to go out to disk.
+         *
+         * A few methods were proposed, but thanks to advice from Perri Boulton (psychology
+         * researcher with a background in statistics), we were able to device a reasonable
+         * approach.
+         *
+         * These are commented in line to help understand the process.
+         */
+
+        /*
+         * Step 1 - we have an index like "idx_eq_member". It has data that looks somewhat
+         * like:
+         *
+         *  key    | idl
+         *  -------+------------
+         *  uuid_a | [1, 2, 3, ...]
+         *  -------+------------
+         *  uuid_b | [4, 5, 6, ...]
+         *
+         * We need to collect this into a single vec of "how long is each idl". Since we have
+         * each idl in the vec, the length of the vec is also the number of keys in the set.
+         * This yields for us:
+         *
+         *   idx_eq_member: [4.0, 5.0, ...]
+         * where each f64 value is the float representation of the length of idl.
+         *
+         * We then assemble these to a map so we have each idxkey and it's associated list
+         * of idl lens.
+         */
+
+        let mut data: HashMap<IdxKey, Vec<f64>> = HashMap::new();
+        self.idl_cache.iter_dirty().for_each(|(k, maybe_idl)| {
```

Comment:
it's part of https://docs.rs/concread/0.2.16/concread/arcache/struct.ARCacheWriteTxn.html#method.iter_dirty which shows you each element in the current working set that is dirty or has been written. Since this is called inside reindex, then "every index" is inside the dirty set at this time, which is why we can do this.

Also because for loops imply .into_iter, not .iter(), and I think .for_each looks prettier.